src/logging.h

```diff
@@ -163,4 +165,84 @@ static inline void LogPrint(const BCLog::LogFlags& category, const Args&... args
     }
 }
 
+
+namespace BCLog {
+
+//! RAII-style object that outputs timing information to logs.
+template <typename TimeType>
+class Timer
+{
+public:
+    //! Log prefix; usually the name of the function this was created in.
+    std::string m_prefix{};
+
+    //! A descriptive message of what is being timed.
+    std::string m_title{};
+
+    //! Forwarded on to LogPrint if specified - has the effect of only
+    //! outputing the timing log when a particular debug= category is specified.
+    BCLog::LogFlags m_log_category{};
+
+    //! If log_category is left as the default, end_msg will log unconditionally
+    //! (instead of being filtered by category).
+    Timer(
+        std::string prefix,
+        std::string end_msg,
+        BCLog::LogFlags log_category = BCLog::LogFlags::ALL) :
+            m_prefix(prefix),
+            m_title(end_msg),
+            m_start_t(GetTime<std::chrono::microseconds>()),
+            m_log_category(log_category) { }
+
+    ~Timer()
+    {
+        this->Log(strprintf("%s completed", m_title));
+    }
+
+    void Log(const std::string& msg)
+    {
+        const std::string full_msg = this->LogMsg(msg);
+
+        if (m_log_category == BCLog::LogFlags::ALL) {
+            LogPrintf(full_msg.c_str()); /* No newline */
```

Comment:
Here and below should this use `LogPrintf("%s", full_msg)` to avoid errors if `msg` contains `%` characters? Using `%s\n` here would also allow dropping the newline in LogMsg and removing the "No newline" lint workarounds.

Also can `.c_str()` be dropped?

Reply:
Ah good suggestions! done